Cabal/src/Distribution/Simple/Build/Monad.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Distribution.Simple.Build.Monad
+  ( BuildM (BuildM)
+  , runBuildM
+  , PreBuildComponentInputs (..)
+
+    -- * A few queries on @'BuildM'@
+  , buildVerbosity
+  , buildWhat
+  , buildComponent
+  , buildCLBI
+  , buildBI
+  , buildLBI
+  , buildCompiler
+  , buildTarget
+
+    -- * Re-exports
+  , BuildingWhat (..)
+  , LocalBuildInfo (..)
+  , TargetInfo (..)
+  , buildingWhatVerbosity
+  , buildingWhatDistPref
+  )
+where
+
+import Control.Monad.Reader
+
+import Distribution.Simple.Compiler
+import Distribution.Simple.Setup (BuildingWhat (..), buildingWhatDistPref, buildingWhatVerbosity)
+import Distribution.Types.BuildInfo
+import Distribution.Types.Component
+import Distribution.Types.ComponentLocalBuildInfo
+import Distribution.Types.LocalBuildInfo
+import Distribution.Types.TargetInfo
+import Distribution.Verbosity
+
+-- | The information required for a build computation (@'BuildM'@)
+-- which is available right before building each component, i.e. the pre-build
+-- component inputs.
+data PreBuildComponentInputs = PreBuildComponentInputs
+  { buildingWhat :: BuildingWhat
+  -- ^ What kind of build are we doing?
+  , localBuildInfo :: LocalBuildInfo
+  -- ^ Information about the package
+  , targetInfo :: TargetInfo
+  -- ^ Information about an individual component
+  }
+
+-- | Computations carried out in the context of building a component (e.g. @'buildAllExtraSources'@)
+newtype BuildM a = BuildM' (ReaderT PreBuildComponentInputs IO a)
+  deriving (Functor, Applicative, Monad, MonadReader PreBuildComponentInputs, MonadIO)
+
+-- Ideally we'd use deriving via ReaderT PreBuildComponentInputs IO, but ghc 8.4 doesn't support it.
```

Comment:
Depending on how you read [the policy](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions), we might not need to support GHC 8.4 anymore. Last 8.4 is 8.4.4, which was released on 14/10/2018, more than 5 years ago.